core/src/main/java/kafka/server/share/ShareFetchUtils.java

```diff
@@ -128,4 +131,17 @@ static long offsetForEarliestTimestamp(TopicIdPartition topicIdPartition, Replic
                 Optional.empty(), true).timestampAndOffsetOpt();
         return timestampAndOffset.isEmpty() ? (long) 0 : timestampAndOffset.get().offset;
     }
+
+    /**
+     * The method is used to get the offset for the latest timestamp for the topic-partition.
+     *
+     * @return The offset for the latest timestamp.
+     */
+    static long offsetForLatestTimestamp(TopicIdPartition topicIdPartition, ReplicaManager replicaManager) {
+        // Isolation level is set to READ_UNCOMMITTED, matching with that used in share fetch requests
+        Option<FileRecords.TimestampAndOffset> timestampAndOffset = replicaManager.fetchOffsetForTimestamp(
+            topicIdPartition.topicPartition(), ListOffsetsRequest.LATEST_TIMESTAMP, new Some<>(IsolationLevel.READ_UNCOMMITTED),
+            Optional.empty(), true).timestampAndOffsetOpt();
+        return timestampAndOffset.isEmpty() ? (long) 0 : timestampAndOffset.get().offset;
```

Comment:
Thanks for the review. We could instead throw an error saying the latest offset not found, or maybe maybe initialize to the earliest offset, in case the latest is not found. What would you suggest here @apoorvmittal10 ?

Reply:
If we send 0 when timestamp is empty but first fetch offset in log is at 10 then will the fetch throw an exception? If yes, then what will happen next? And in waht cases the the timestamp can be empty from replica manager?

Reply:
I think there is the possibility of an exception in the slightly unusual situation of a new leader which is lagging behind the current epoch. While I'm sure we're not going to bump into that right now, it is a real situation in the replica manager I think and the interface here needs to be flexible enough for an exception and we need to have a plan for what to do if the exception occurs. @mumrah can you help here please? I guess we'd want to do some kind of retry, which is perhaps best achieved by aborting the share-partition initialisation and letting the next attempt to initialize the share-partition have another try. The edges of the replication behaviour are a bit challenging I think.

Reply:
Hi @apoorvmittal10 , when `replicaManager.fetchOffsetForTimestamp` is called with `LATEST_TIMESTAMP`, as per the current code, only 2 things are possible, either it will throw an error or it will return a populated response (not a None object, which is possible when the same method is called with `EARLIEST_TIMESTAMP`). Now the error handling is there for when an error is thrown, but apart from that I don't suppose we need any other checks for whether the offset value is populated or not. But, if we want to have this check in place, I don't think we can assume any other value in place of this. The only right way would be to fail the share partition initialization. Another point to note is when `offsetForEarliestTimestamp` is called with `EARLIEST_TIMESTAMP`, the current code would return 0 if the returned value from `replicaManager.fetchOffsetForTimestamp` is None, which again would be a wrong assumption in case the start offset of the log is beyond 0. I believe the correct way is to fail the share partition initialization here as well.